fmpz/comb.c

```diff
@@ -0,0 +1,351 @@
+
+    CT->A = _fmpz_vec_init(CT->Alen);
+    CT->T = _fmpz_vec_init(CT->Tlen);
+}
+
+
+void fmpz_comb_temp_clear(fmpz_comb_temp_t CT)
+{
+    _fmpz_vec_clear(CT->A, CT->Alen);
+    _fmpz_vec_clear(CT->T, CT->Tlen);
+}
+
+
+void fmpz_comb_clear(fmpz_comb_t C)
+{
+    flint_free(C->step);
+    flint_free(C->packed_multipliers);
+    flint_free(C->crt_lu);
+    flint_free(C->mod_lu);
+    flint_free(C->crt_offsets);
+    flint_free(C->mod_offsets);
+    fmpz_multi_CRT_clear(C->crt_P);
+    fmpz_multi_mod_clear(C->mod_P);
+}
+
+
+void fmpz_comb_init(fmpz_comb_t C, mp_srcptr m, slong len)
+{
+    int success;
+    slong l, i, j, k;
+    ulong tt, mm, mt;
+    fmpz_poly_t M, Mm; /* only used for resizable fmpz array convenience */
+
+    if (len < 1)
+        flint_throw(FLINT_ERROR, "fmpz_comb_init");
+
+    fmpz_poly_init(Mm);
+    fmpz_poly_init(M);
+
+    fmpz_multi_CRT_init(C->crt_P);
+    C->packed_multipliers = NULL;
+    C->step = NULL;
+    C->crt_lu = NULL;
+    C->crt_lu_alloc = 0;
+    C->crt_offsets = NULL;
+    C->crt_offsets_alloc = 0;
+
+    fmpz_multi_mod_init(C->mod_P);
+    C->mod_lu = NULL;
+    C->mod_lu_alloc = 0;
+    C->mod_offsets = NULL;
+    C->mod_offsets_alloc = 0;
+
+    C->num_primes = len;
+
+    /* set up crt */
+
+    for (l = 0, i = 0, k = 0; l < len; k++)
+    {
+        if (k + 1 >= C->crt_offsets_alloc)
+        {
+            C->crt_offsets_alloc = FLINT_MAX(k + 1, C->crt_offsets_alloc*3/2);
+            C->crt_offsets = FLINT_ARRAY_REALLOC(C->crt_offsets,
+                                                  C->crt_offsets_alloc, slong);
+        }
+
+        fmpz_poly_fit_length(M, k + 1);
+        fmpz_one(M->coeffs + k);
+
+        for (j = i;
+             l < len && fmpz_size(M->coeffs + k) <= FMPZ_CRT_UI_CUTOFF;
+             j++)
+        {
+            if (j + 1 >= C->crt_lu_alloc)
+            {
+                C->crt_lu_alloc = FLINT_MAX(j + 1, C->crt_lu_alloc*3/2);
+                C->crt_lu = FLINT_ARRAY_REALLOC(C->crt_lu, C->crt_lu_alloc,
+                                                                crt_lut_entry);
+            }
+
+            C->crt_lu[j].i0 = 0;
+            C->crt_lu[j].i1 = 0;
+            C->crt_lu[j].i2 = 0;
+
+            mm = m[l];
+            if (l + 1 < len && !n_mul_checked(&mt, mm, m[l+1]))
+            {
+                mm = mt;
+                if (l + 2 < len && !n_mul_checked(&mt, mm, m[l+2]))
+                {
+                    mm = mt;
+                    success = (1 == n_gcdinv(&C->crt_lu[j].i0,
+                                          m[l+1]*m[l+2] % m[l+0], m[l+0])) &&
+                              (1 == n_gcdinv(&C->crt_lu[j].i1,
+                                          m[l+0]*m[l+2] % m[l+1], m[l+1])) &&
+                              (1 == n_gcdinv(&C->crt_lu[j].i2,
+                                          m[l+0]*m[l+1] % m[l+2], m[l+2]));
+                    if (!success)
+                        flint_throw(FLINT_ERROR, "fmpz_comb_init");
```

Comment:
Any chance of a more descriptive error? And below.

Reply:
yup